### DISCUSSION TITLE
ControlNet Disappeared from my UI

### DISCUSSION THREAD

**Discussion by KazFoxsen** [Mikubill/sd-webui-controlnet#687] [Q&A] [answered]
It used to work fine, but I've been trying things to fix a CUDA error when training and I don't know if I've broken something with ControlNet.

I've tried deleting ControlNet and reinstalling, but that didn't work. The README file for ControlNet mentions this: Upgrade gradio if any ui issues occured: `pip install gradio==3.16.2`
I tried that, but restarting Auto1111 still says Gradio is 3.23.0.

The command prompts gives these errors:
`Error loading script: api.py, Error loading script: controlnet.py`, `Error loading script: external_code.py`, and `ImportError: cannot import name 'api' from 'modules.api' (unknown location)`
How can I fix this?

**Comment by ljleb** [comment]
Really not sure about this. Which version of python are you using?
Also you need to activate your venv before installing anything, otherwise it will install in your global package repository and won't change the local version used by the webui.

**Reply by KazFoxsen** [reply]
My Python is 3.10.6. which I guess is the only version that Automatic 1111 works with? When I install things, I do use the venv\scripts folder and type "activate" in the command prompt before installing stuff.

**Comment by KazFoxsen** [comment, marked as answer]
I wonder if this has anything to do with the stuff in this thread? I still don't know how to fix it though. I'm not a dev, so it's over my head.
https://github.com/Mikubill/sd-webui-controlnet/issues/567

**Reply by ljleb** [reply]
No it does not seem related at a first glance. It seems any module controlnet tries to load cannot be found. Can you share the full stack trace from your console? As the location where the error happens matters. As `controlnet.py` is among the failed modules, I expect this could be caused by another extension, however I am not sure.

**Reply by KazFoxsen** [reply]
Here's what my console says:

```
error: Your local changes to the following files would be overwritten by merge:
        README.md
        extensions-builtin/Lora/lora.py
        javascript/aspectRatioOverlay.js
        javascript/hints.js
        javascript/imageviewer.js
        javascript/ui.js
Please commit your changes or stash them before you merge.
Aborting
Updating 4c1ad743..22bcc7be
venv "D:\stable-diffusion-webui\venv\Scripts\Python.exe"
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Commit hash: 4c1ad743e3baf1246db0711aa0107debf036a12b
Installing requirements for Web UI

Launching Web UI with arguments: --opt-sdp-attention --medvram
No module 'xformers'. Proceeding without it.
D:\stable-diffusion-webui\venv\lib\site-packages\torchvision\transforms\functional_tensor.py:5: UserWarning: The torchvision.transforms.functional_tensor module is deprecated in 0.15 and will be **removed in 0.17**. Please don't rely on it. You probably just need to use APIs in torchvision.transforms.functional or in torchvision.transforms.v2.functional.
  warnings.warn(
Error loading script: api.py
Traceback (most recent call last):
  File "D:\stable-diffusion-webui\modules\scripts.py", line 256, in load_scripts
    script_module = script_loading.load_module(scriptfile.path)
  File "D:\stable-diffusion-webui\modules\script_loading.py", line 11, in load_module
    module_spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\api.py", line 13, in <module>
    from modules.api import api
ImportError: cannot import name 'api' from 'modules.api' (unknown location)

Error loading script: controlnet.py
Traceback (most recent call last):
  File "D:\stable-diffusion-webui\modules\scripts.py", line 256, in load_scripts
    script_module = script_loading.load_module(scriptfile.path)
  File "D:\stable-diffusion-webui\modules\script_loading.py", line 11, in load_module
    module_spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 20, in <module>
    from scripts import external_code, global_state
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\external_code.py", line 7, in <module>
    from modules.api import api
ImportError: cannot import name 'api' from 'modules.api' (unknown location)

Error loading script: external_code.py
Traceback (most recent call last):
  File "D:\stable-diffusion-webui\modules\scripts.py", line 256, in load_scripts
    script_module = script_loading.load_module(scriptfile.path)
  File "D:\stable-diffusion-webui\modules\script_loading.py", line 11, in load_module
    module_spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "D:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\external_code.py", line 7, in <module>
    from modules.api import api
ImportError: cannot import name 'api' from 'modules.api' (unknown location)

Loading weights [ad2a33c361] from D:\stable-diffusion-webui\models\Stable-diffusion\sd2-v2-1_768-ema-pruned.ckpt
Creating model from config: D:\stable-diffusion-webui\repositories\stable-diffusion-stability-ai\configs\stable-diffusion\v2-inference-v.yaml
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
Applying scaled dot product cross attention optimization.
Textual inversion embeddings loaded(1): Nudge
Model loaded in 95.6s (load weights from disk: 48.7s, find config: 8.0s, create model: 0.5s, apply weights to model: 8.9s, apply half(): 3.1s, load VAE: 0.2s, move model to device: 0.4s, load textual inversion embeddings: 25.6s).
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 173.4s (import torch: 37.5s, import gradio: 5.6s, import ldm: 1.2s, other imports: 10.2s, list extensions: 1.4s, list SD models: 0.1s, setup codeformer: 0.6s, load scripts: 2.3s, load SD checkpoint: 96.0s, create ui: 17.7s, gradio launch: 0.6s).
```

**Reply by ljleb** [reply]
You seem to have local changes in your webui, + you seem to try to update to the latest version of the webui every time you start it. I do not recommend to have any of these if you are not a developer.

If you did not intend to have local changes, you should consider resetting your git repo. To do this, enter `git reset --hard HEAD` in the command line. It will keep your `webui-user.bat` untouched. After this, you may want to consider updating to the latest version of the webui as, AFAIK the latest commit fixes a couple things.

If updating to the latest version does not fix the issue let me know, I'll see if I can help you further.

**Reply by KazFoxsen** [reply]
I’m not sure what “local changes” are in this context. The tutorial I followed a while back to install Auto1111 instructed to add “git pull” to the webui-user.bat file to automatically update, but I’ve since read in forums that it’s bad to do that. I’m not very familiar with Git and have found it confusing (I’m not a dev).
I activated the venv\Scripts folder and entered the command `git reset --hard HEAD` (at first, I thought it was something you add to the BAT file’s command line arguments, but it didn’t recognize the command)

It said this:
`HEAD is now at 4c1ad743 for img2img, use None as upscaler instead of erroring out if the desired upscaler is not found`
I don't know what that means, but when I started up Auto1111, ControlNet was back!

From now on, how do I update Auto1111?

**Reply by ljleb** [reply]
Nice! To update a1111 webui, open a terminal and type `git pull origin master`. You may want alternatively to use the v1.0.0-pre zip: https://github.com/AUTOMATIC1111/stable-diffusion-webui/releases/tag/v1.0.0-pre (the one named `sd.webui.zip`)

It includes 2 scripts: `update.bat` and `run.bat`. to update, simply close the webui and click on `update.bat`. To run, simply click on `run.bat`. The configuration process is otherwise the same, the full webui repo is located under `webui` in the zip.

**Reply by KazFoxsen** [reply]
Thank you!

**Comment by KazFoxsen** [comment]
If this helps anyone else too, double check your BAT file after the git reset command. I had to put my commandline args back into it.

**Comment by DryIceX** [comment]
I still couldn't get this to work. Gonna do a full re-install without the git pull command line and hopefully won't have any issues after that.
